util/crc32c_arm64.cc

```diff
@@ -19,35 +19,82 @@ uint32_t crc32c_runtime_check(void) {
 
 uint32_t crc32c_arm64(uint32_t crc, unsigned char const *data,
                              unsigned len) {
-  const uint8_t *buf1;
-  const uint16_t *buf2;
-  const uint32_t *buf4;
-  const uint64_t *buf8;
+  const uint8_t *buf8;
+  const uint64_t *buf64 = (uint64_t *)data;
+  int length = (int)len;
+  crc ^= 0xffffffff;
 
-  int64_t length = (int64_t)len;
+#ifdef HAVE_ARM64_CRYPTO
+  /* Crc32c Parallel computation
+   *   Algorithm comes from Intel whitepaper:
+   *   crc-iscsi-polynomial-crc32-instruction-paper
+   *
+   * Input data is divided into three equal-sized blocks
+   *   Three parallel blocks (crc0, crc1, crc2) for 1024 Bytes
+   *   One Block: 42(BLK_LENGTH) * 8(step length: crc32c_u64) bytes
+   */
+  #define BLK_LENGTH 42
+  while (length >= 1024) {
+    uint64_t t0, t1;
+    uint32_t crc0 = 0, crc1 = 0, crc2 = 0;
 
-  crc ^= 0xffffffff;
-  buf8 = (const uint64_t *)data;
-  while ((length -= sizeof(uint64_t)) >= 0) {
-    crc = __crc32cd(crc, *buf8++);
+    /* Parallel Param:
+     *   k0 = CRC32(x ^ (42 * 8 * 8 * 2 - 1));
+     *   k1 = CRC32(x ^ (42 * 8 * 8 - 1));
+     */
+    uint32_t k0 = 0xe417f38a, k1 = 0x8f158014;
+
+    /* First 8 bytei for better pipelining */
+    crc0 = crc32c_u64(crc, *buf64++);
+
+    /* 3 blocks crc32c parallel computation
+     *
+     * 42 * 8 * 3 = 1008 (bytes)
+     */
+    for (int i = 0; i < BLK_LENGTH; i++, buf64++) {
+      crc0 = crc32c_u64(crc0, *buf64);
+      crc1 = crc32c_u64(crc1, *(buf64 + BLK_LENGTH));
+      crc2 = crc32c_u64(crc2, *(buf64 + (BLK_LENGTH * 2)));
+    }
+    buf64 += (BLK_LENGTH * 2);
+
+    /* Last 8 bytes */
+    crc = crc32c_u64(crc2, *buf64++);
```

Comment:
Really thanks for your comments.

The first 8 bytes would be calculated at Line47~48: 
 ```
   /* First 8 bytei for better pipelining */
    crc0 = crc32c_u64(crc, *buf64++);
```

Reply:
Interesting...
This is a slightly different approach from the 3-way CRC used in the crc32c.cc. The downside of this approach is that, it doesn't apply 3-way CRC to the remainder of bytes after the 1KB loop. Since RocksDB's block size is not aligned to 1KB, this approach is probably worse than the one in crc32c. It's still a better one than what is already there. I'm going to merge it now.